internal/cmd/root.go

```diff
@@ -1,83 +1,139 @@
 package cmd
 
 import (
+	_ "embed"
 	"fmt"
 	"os"
 	"strings"
 
+	"github.com/gobuffalo/cli/internal/cmd/build"
+	"github.com/gobuffalo/cli/internal/cmd/destroy"
+	"github.com/gobuffalo/cli/internal/cmd/dev"
+	"github.com/gobuffalo/cli/internal/cmd/fix"
+	"github.com/gobuffalo/cli/internal/cmd/generate"
+	"github.com/gobuffalo/cli/internal/cmd/info"
+	"github.com/gobuffalo/cli/internal/cmd/new"
+	cmdplugins "github.com/gobuffalo/cli/internal/cmd/plugins"
+	"github.com/gobuffalo/cli/internal/cmd/routes"
+	"github.com/gobuffalo/cli/internal/cmd/setup"
+	"github.com/gobuffalo/cli/internal/cmd/task"
+	"github.com/gobuffalo/cli/internal/cmd/test"
+	"github.com/gobuffalo/cli/internal/cmd/version"
 	"github.com/gobuffalo/cli/internal/plugins"
 	"github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 )
 
-var anywhereCommands = []string{"new", "version", "info", "help"}
-
-// RootCmd is the hook for all of the other commands in the buffalo binary.
-var RootCmd = &cobra.Command{
-	SilenceErrors: true,
-	Use:           "buffalo",
-	Short:         "Build Buffalo applications with ease",
-	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
-		if err := plugins.Load(); err != nil {
-			return err
-		}
-
-		isFreeCommand := false
-		for _, freeCmd := range anywhereCommands {
-			if freeCmd == cmd.Name() {
-				isFreeCommand = true
-			}
-		}
-
-		if isFreeCommand {
-			return nil
-		}
+const (
+	dbNotFound  = `unknown command "db"`
+	popNotFound = `unknown command "pop"`
+)
 
-		if !insideBuffaloProject() {
-			return fmt.Errorf("you need to be inside your buffalo project path to run this command")
-		}
+var (
+	anywhereCommands = []string{"new", "version", "info", "help"}
 
-		return nil
-	},
-}
+	//go:embed popinstructions.txt
+	popInstallInstructions string
+)
 
 // Execute adds all child commands to the root command sets flags appropriately.
 // This is called by main.main(). It only needs to happen once to the rootCmd.
 func Execute() {
-	if err := RootCmd.Execute(); err != nil {
-		if strings.Contains(err.Error(), dbNotFound) || strings.Contains(err.Error(), popNotFound) {
-			logrus.Errorf(popInstallInstructions)
-			os.Exit(-1)
-		}
-		logrus.Errorf("Error: %s", err)
-		if strings.Contains(err.Error(), dbNotFound) || strings.Contains(err.Error(), popNotFound) {
-			fmt.Println(popInstallInstructions)
-			os.Exit(-1)
-		}
+	rootCmd := cmd()
+	err := rootCmd.Execute()
+	if err == nil {
+		return
+	}
+
+	if strings.Contains(err.Error(), dbNotFound) || strings.Contains(err.Error(), popNotFound) {
+		logrus.Errorf(popInstallInstructions)
 		os.Exit(-1)
 	}
+	logrus.Errorf("Error: %s", err)
+	if strings.Contains(err.Error(), dbNotFound) || strings.Contains(err.Error(), popNotFound) {
```

Comment:
why the same if twice?

Reply:
Not sure, and there are no tests about this behavior anywhere :D